students/alex_whitty/lesson4/trigrams.py

```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python3
+
+with open('double_take.txt', 'r') as f:
+
+    for line in f:
+        print(line, end='')
+
+
+
+
+trigrams = {("When", "the"): ['Travelers'],
+            ("the", "Travelers"): ["from"],
+            ("Travelers", "from"): ["Outer"],
+            ("from", "Outer"): ["space"],
+            ("Outer", "Space"): ["dug"],
+            ("Space", "dug"): ["into"],
+            ("dug", "into"): ["the"],
+            ("into", "the"): ["pile"],
+            }
+
+
+
+def build_trigrams(words):
+
+    trigrams = {}
+
+    for i in range(len(words) - 2):
+
+
+        pair = tuple(words[i:i + 2])
+        follower = words[i + 2]
+
+        if pair not in trigrams:
+
+            trigrams[pair] = [follower]
+        else:
+            trigrams[pair].append(follower)
+
+
+    return trigrams
+
+# build up the dict here!
+
+def build_dict(words):
+
+    import random
+
+    random.choice(words)
+
+    new_text = 'from outer'
+
+    if words == {('from', 'outer'): ['Space']}:
+        new_text.append("Space")
+        return new_text
+
```

Comment:
This one is trickier, you need to do two different random selections. You will first select a random key from the trigrams dictionary, those will be the first two words of the new text. After that, randomly select one word from the list associated the key, that will be the third word of the new text. Now, you need to take the second word of that first key and the word you picked from the list and form a tuple, which will be the next key. You need to repeat this process either until you create a key that is not in the dictionary or you meet some other stop condition of your own choosing. The code below starts the process:

```
def build_dict(words):
    starter = random.choice(list(words.keys()))
    print(starter)
    # Start the text with the random key
    # now turned into a string
    new_text = ' '.join(starter)

    # Now, pick a random word from the list
    # associated to the key
    key_value = words[starter]
    next_word = random.choice(key_value)

    # Add next_word to new_text
    new_text += ' ' + next_word
    print(new_text)

    # Form the next round, using the second
    # word of the starter and next_word
    # and repeat the process
```